Problem set 2: Aggregation, reshaping, and validation

Due: 11:59pm on Friday, September 13 by uploading to Brightspace

Your name:

1. For this problem set, use any of the following datasets:

* [NOAA](https://www.ncei.noaa.gov/products) weather and climate data
* [World Happiness Report](https://worldhappiness.report/ed/2021/#appendices-and-data)
* [NASA data](https://data.nasa.gov/)
* [US Census data](https://data.census.gov/)
* [WHO COVID data](https://covid19.who.int/)
* [Netflix data](https://www.kaggle.com/datasets/ashishgup/netflix-rotten-tomatoes-metacritic-imdb)
* [Nashville open data](https://data.nashville.gov/)
* [Yelp reviews data](https://www.yelp.com/dataset)

In [19]:
# Read in the data
import pandas as pd
df = pd.read_excel('./WHR21_Data_Mortality.xlsx')

# Do some investigative EDA to identify the structure of the data and its content. Then write a brief summary of the dataset in your own words.
print(df.info())

# Rename columns for easier access
df = df.rename(columns={
    "Country name": "country_name",
    "Population 2020": "population_2020",
    "Population 2019": "population_2019",
    "COVID-19 deaths per 100,000 population in 2020": "covid19_deaths_per_100k_2020",
    "Median age": "median_age",
    "Island": "island",
    "Index of exposure to COVID-19  infections in other countries as of March 31": "exposure_index_march31",
    "Log of average distance to SARS countries": "log_avg_distance_sars_countries",
    "WHO Western Pacific Region": "who_western_pacific_region",
    "Female head of government": "female_head_of_government",
    "Index of institutional trust": "institutional_trust_index",
    "Gini coefficient of income": "gini_coefficient",
    "All-cause death count, 2017": "all_cause_deaths_2017",
    "All-cause death count, 2018": "all_cause_deaths_2018",
    "All-cause death count, 2019": "all_cause_deaths_2019",
    "All-cause death count, 2020": "all_cause_deaths_2020",
    "Excess deaths in 2020 per 100,000 population, relative to 2017-2019 average": "excess_deaths_per_100k_2020"
})
print(df.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 166 entries, 0 to 165
Data columns (total 17 columns):
 #   Column                                                                       Non-Null Count  Dtype  
---  ------                                                                       --------------  -----  
 0   Country name                                                                 166 non-null    object 
 1   Population 2020                                                              164 non-null    float64
 2   Population 2019                                                              164 non-null    float64
 3   COVID-19 deaths per 100,000 population in 2020                               163 non-null    float64
 4   Median age                                                                   163 non-null    float64
 5   Island                                                                       166 non-null    int64  
 6   Index of exposure to COVID-19  infections 

In [20]:
display(df.sample(10))

country_name  population_2020  population_2019  \
14   Netherlands       17134873.0       17332850.0   
9      Indonesia      273523621.0      270625568.0   
21       Romania       19237682.0       19356544.0   
64        Zambia       18383956.0       17861030.0   
25          Iran       83992953.0       82913906.0   
33        Mexico      128932753.0      127575529.0   
76      Cameroon       26545864.0       25876380.0   
63       Senegal       16743930.0       16296364.0   
4   Saudi Arabia       34813867.0       34268528.0   
60    Mozambique       31255435.0       30366036.0   

    covid19_deaths_per_100k_2020  median_age  island  exposure_index_march31  \
14                     67.260498   43.200001       0                5.573827   
9                       8.093599   29.299999       1                0.575605   
21                     81.958900   43.000000       0                2.516937   
64                      2.110500   17.700001       0                0.666667   
25                     65.747200   32.400002       0                1.706699   
33                     97.575699   29.299999       0                0.847846   
76                      1.687600   18.799999       0                0.919348   
63                      2.448600   18.700001       0                1.061633   
4                      17.875101   31.900000       0                1.250483   
60                      0.531100   17.700001       0                0.598771   

    log_avg_distance_sars_countries  who_western_pacific_region  \
14                         9.059903                           0   
9                          8.577746                           0   
21                         8.980768                           0   
64                         9.251140                           0   
25                         8.809122                           0   
33                         9.433811                           0   
76                         9.275795                           0   
63                         9.390359                           0   
4                          8.896798                           0   
60                         9.272958                           0   

    female_head_of_government  institutional_trust_index  gini_coefficient  \
14                          0                   0.522208         27.000000   
9                           0                   0.560791         37.477500   
21                          0                   0.100040         35.100000   
64                          0                   0.321515         57.140000   
25                          0                   0.367223         39.970000   
33                          0                   0.128877         47.500000   
76                          0                   0.281469         46.640000   
63                          0                   0.355424         40.290000   
4                           0                   0.650619         45.900000   
60                          0                   0.472666         49.416667   

    all_cause_deaths_2017  all_cause_deaths_2018  all_cause_deaths_2019  \
14               149805.0               152991.0               151547.0   
9                     NaN                    NaN                    NaN   
21                    NaN                    NaN                    NaN   
64                    NaN                    NaN                    NaN   
25                    NaN                    NaN                    NaN   
33               701160.0               718206.0               728889.0   
76                    NaN                    NaN                    NaN   
63                    NaN                    NaN                    NaN   
4                     NaN                    NaN                    NaN   
60                    NaN                    NaN                    NaN   

    all_cause_deaths_2020  excess_deaths_per_100k_2020  
14               171175.0                   114.

### Summary
*This dataset, from World Hapiness Report, contains information on 166 countries, including their demographics (such as population size and age distribution), pandemic data (COVID-19 deaths and excess mortality), structural features (island status, exposure to international spread), and social-economic factors (institutional trust, income inequality, female leadership) in 2020. Most data are well documented, while records of all-cause mortality from previous years are less complete.*


2. Aggregate and reshape

In [21]:
# Reshape the data in some way that changes the unit of analysis. Convert it into an aggregated summary table that provides new information from that table
group_by_island = df.groupby("island").agg({
    "population_2020": "mean",
    "median_age": "mean",
    "covid19_deaths_per_100k_2020": "mean",
    "excess_deaths_per_100k_2020": "mean",
    "institutional_trust_index": "mean"
}).reset_index()

group_by_island
# Write 3-4 sentences summarizing what new insight we get from the aggregated table


island  population_2020  median_age  covid19_deaths_per_100k_2020  \
0       0     4.947931e+07   30.101418                     33.172511   
1       1     3.234825e+07   34.295455                     14.883223   

   excess_deaths_per_100k_2020  institutional_trust_index  
0                   119.959358                   0.308743  
1                    43.188946                   0.369632

### Summary

*From the aggregated data, we see island nations tend to have smaller populations and  older median ages compared to non-island countries. They had lower COVID-19 deaths and excess mortality rates, potentially due to their geographic isolation. Island countries also show somewhat higher institutional trust.*


3. Validation and missingness


In [22]:
# Choose one numeric column from either one of the original tables or your newly merged table. Convert a random 5% sample of values to NaN. Make sure you set a seed first so the dropped values are always the same
import numpy as np

np.random.seed(42)

df_num = df.copy()

random_indices = np.random.choice(df_num.index, size=int(0.05 * len(df_num)), replace=False)
df_num.loc[random_indices, "COVID-19 deaths per 100,000 population in 2020"] = np.nan

# Calculate the percent of missing values in each column

missing_percent = df_num.isnull().mean() * 100
print("Percent of missing values in each column:")
print(missing_percent)

# Find the mean of the numeric column

mean_value = df_num["COVID-19 deaths per 100,000 population in 2020"].mean()
print(f"Mean of 'COVID-19 deaths per 100,000 population in 2020': {mean_value}")

# Use listwise deletion and find the mean again

mean_value_listwise = df_num["COVID-19 deaths per 100,000 population in 2020"].dropna().mean()
print(f"Mean of 'COVID-19 deaths per 100,000 population in 2020' (listwise deletion): {mean_value_listwise}")

# Use mean imputation and find the mean again

df_num["COVID-19 deaths per 100,000 population in 2020"].fillna(mean_value, inplace=True)
mean_value_imputed = df_num["COVID-19 deaths per 100,000 population in 2020"].mean()
print(f"Mean of 'COVID-19 deaths per 100,000 population in 2020' (mean imputation): {mean_value_imputed}")

Percent of missing values in each column:
country_name                                        0.000000
population_2020                                     1.204819
population_2019                                     1.204819
covid19_deaths_per_100k_2020                        1.807229
median_age                                          1.807229
island                                              0.000000
exposure_index_march31                              1.204819
log_avg_distance_sars_countries                     1.204819
who_western_pacific_region                          0.000000
female_head_of_government                           0.000000
institutional_trust_index                           0.602410
gini_coefficient                                    1.204819
all_cause_deaths_2017                              62.048193
all_cause_deaths_2018                              61.445783
all_cause_deaths_2019                              61.445783
all_cause_deaths_2020                      

C:\temp\ipykernel_5284\2331257004.py:29: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_num["COVID-19 deaths per 100,000 population in 2020"].fillna(mean_value, inplace=True)


In [23]:
# Use a more advanced form of imputation and find the mean again
df_num = df.copy()
random_indices = np.random.choice(df_num.index, size=int(0.05 * len(df_num)), replace=False)
df_num.loc[random_indices, "COVID-19 deaths per 100,000 population in 2020"] = np.nan

# Median imputation
median_value = df_num["COVID-19 deaths per 100,000 population in 2020"].median()
df_num_median = df_num.copy()
df_num_median["COVID-19 deaths per 100,000 population in 2020"].fillna(median_value, inplace=True)

mean_value_median = df_num_median["COVID-19 deaths per 100,000 population in 2020"].mean()
print(f"Mean of 'COVID-19 deaths per 100,000 population in 2020' (median imputation): {mean_value_median}")


Mean of 'COVID-19 deaths per 100,000 population in 2020' (median imputation): nan


C:\temp\ipykernel_5284\4147809403.py:9: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_num_median["COVID-19 deaths per 100,000 population in 2020"].fillna(median_value, inplace=True)


### Compare these 4 values and in 3-4 sentences decribe that it tells us about imputation and missing data

*When comparing the four approaches, the values are close but not equal. Listwise deletion reduces the sample size and can slightly bias the result if the missing values are not random. Mean imputation keeps the mean stable but reduces the variance. Median imputation is less sensitive to outliers.*

In [24]:
# Repeat the above steps with a factor/string column, this time doing mode imputation instead of mean
df_cat = df.copy()
np.random.seed(42)

random_indices = np.random.choice(df_cat.index, size=int(0.05 * len(df_cat)), replace=False)
df_cat.loc[random_indices, "female_head_of_government"] = np.nan

missing_percent_cat = df_cat.isnull().mean() * 100
print("Percent of missing values in each column:")
print(missing_percent_cat)

# Original mode
true_mode = df["female_head_of_government"].mode()[0]

# Listwise deletion
listwise_mode = df_cat["female_head_of_government"].dropna().mode()[0]

# Mode imputation
mode_value = df_cat["female_head_of_government"].mode()[0]
df_cat_mode = df_cat.copy()
df_cat_mode["female_head_of_government"].fillna(mode_value, inplace=True)
mode_imputed = df_cat_mode["female_head_of_government"].mode()[0]

print(f"Original mode: {true_mode}")
print(f"Mode after listwise deletion: {listwise_mode}")
print(f"Mode after mode imputation: {mode_imputed}")

Percent of missing values in each column:
country_name                        0.000000
population_2020                     1.204819
population_2019                     1.204819
covid19_deaths_per_100k_2020        1.807229
median_age                          1.807229
island                              0.000000
exposure_index_march31              1.204819
log_avg_distance_sars_countries     1.204819
who_western_pacific_region          0.000000
female_head_of_government           4.819277
institutional_trust_index           0.602410
gini_coefficient                    1.204819
all_cause_deaths_2017              62.048193
all_cause_deaths_2018              61.445783
all_cause_deaths_2019              61.445783
all_cause_deaths_2020              60.843373
excess_deaths_per_100k_2020        61.445783
dtype: float64
Original mode: 0
Mode after listwise deletion: 0.0
Mode after mode imputation: 0.0


C:\temp\ipykernel_5284\1103400183.py:21: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_cat_mode["female_head_of_government"].fillna(mode_value, inplace=True)
